spring-cloud-huawei-canary-common/src/main/java/org/springframework/cloud/canary/common/distribute/AbstractCanaryDistributer.java

```diff
@@ -0,0 +1,139 @@
+        this.getVersion = getVersion;
+        this.getServerName = getServerName;
+        this.getProperties = getProperties;
+    }
+
+    public List<T> getFiltedServer(Map<TagItem, List<T>> allServer,
+                                   PolicyRuleItem rule, String targetServiceName) {
+        return allServer
+                .get(CanaryRuleCache.getServiceInfoCacheMap().get(targetServiceName).getNextInvokeVersion(rule));
+    }
+
+    /**
+     * 1.过滤targetService
+     * 2.返回按照version和tags分配list
+     *
+     * @param serviceName
+     * @param list
+     * @return
+     */
+    private Map<TagItem, List<T>> getDistributList(String serviceName,
+                                                   List<T> list,
+                                                   PolicyRuleItem invokeRule) {
+        String latestV = CanaryRuleCache.getServiceInfoCacheMap().get(serviceName).getLatestVersionTag().getVersion();
+        Map<TagItem, List<T>> versionServerMap = new HashMap<>();
+        invokeRule.getRoute().forEach(a ->
+                versionServerMap.put(a.getTagitem(), new ArrayList<>())
+        );
+        for (T server : list) {
+            //获得目标服务
+            E ms = getIns.apply(server);
+            if (getServerName.apply(ms).equals(serviceName)) {
+                //最多匹配原则
+                TagItem tagitem = new TagItem(getVersion.apply(ms), getProperties.apply(ms));
+                TagItem targetTag = null;
+                int maxMatch = 0;
+                for (Map.Entry<TagItem, List<T>> entry : versionServerMap.entrySet()) {
+                    int nowMatch = entry.getKey().matchNum(tagitem);
+                    if (nowMatch > maxMatch) {
+                        maxMatch = nowMatch;
+                        targetTag = entry.getKey();
+                    }
+                }
+                if (invokeRule.isWeightLess() && getVersion.apply(ms).equals(latestV)) {
+                    versionServerMap.get(invokeRule.getRoute().get(invokeRule.getRoute().size() - 1).getTagitem()).add(server);
+                }
+                versionServerMap.get(targetTag).add(server);
+            }
+        }
+        return versionServerMap;
+    }
+
+    public void initVer(String serviceName, List<T> list) {
+        if (CanaryRuleCache.getServiceInfoCacheMap().get(serviceName).getLatestVersionTag() != null) {
+            return;
+        }
+        String latestVersion = null;
+        for (T server : list) {
+            E ms = getIns.apply(server);
+            if (getServerName.apply(ms).equals(serviceName)) {
+                if (latestVersion == null || VersionCompareUtil.compareVersion(latestVersion, getVersion.apply(ms)) == -1) {
+                    latestVersion = getVersion.apply(ms);
+                }
+            }
+        }
+        Map<String, String> temMap = new HashMap<>();
+        temMap.put("version", latestVersion);
+        TagItem tagitem = new TagItem(temMap.get("version"), temMap);
```

Comment:
map can be removed

Reply:
how to do?

Reply:
TagItem tagitem = new TagItem("version",latestVersion);

Can be handled in the TagItem constructor, because there may be another place with the same logic

        TagItem tagitem = new TagItem(temMap);
or  only pass one parameter map, otherwise the code doesn't look clean

"version" should be extracted

Reply:
done